File: WeekendContest/LeetCode_2477.cpp
// 2477. Minimum Fuel Cost to Report to the Capital

class Solution {
public:    
    long long minimumFuelCost(vector<vector<int>>& roads, int seats) {
        vector<vector<int>> e(roads.size() + 1); // prepare the edges 
        for (auto &r : roads) 
            e[r[0]].push_back(r[1]),e[r[1]].push_back(r[0]);
        
        long long ans = 0;
        function<int(int,int,int)> dfs = [&](int i, int pre, int depth){
            int riders = 1; // current node
            for (int j : e[i])
                if (j != pre)
                    riders += dfs(j, i, depth + 1);
            if (depth == 0) return 0;
            ans += (riders / seats) * depth + (riders % seats ? 1 : 0); // Nice here
            return riders % seats;
        };
        
        dfs(0, -1, 0);
        return ans;
    }
};

File: DailyChallenges/Scala/LeetCode_2477.scala
object Solution {
    def minimumFuelCost(roads: Array[Array[Int]], seats: Int): Long = {
        val n = roads.length+1
        val e = Array.fill[List[Int]](n)(Nil)
        for (a <- roads){
            e(a(0)) = a(1) :: e(a(0))
            e(a(1)) = a(0) :: e(a(1))
        }

        var ans:Long = 0L
        def dfs(i: Int, pre: Int, depth: Int): Int = {
            var riders = 1
            for (j <- e(i) if j != pre)
                riders += dfs(j, i, depth + 1)
            if (depth == 0) 0 else {
                ans += (riders / seats) * depth + (if (riders % seats == 0) 0 else 1)
                riders % seats
            }
        }

        dfs(0, -1, 0)
        ans
    }
}
